### ISSUE TITLE
Unable to save Github Credentials on Linux

### ISSUE THREAD

**Issue by classabbyamp** [microsoft/vscode#120392]
- VS Code Version: 1.54.3
- OS Version: Arch Linux x64 kernel 5.11.10-arch1-1

Steps to Reproduce:

1. Click the person icon in the bottom right
2. Click "Sign in to Settings Sync"
3. Select "Sign in with GitHub"
4. Browser opens, does Oauth flow, and returns to VS Code
5. Error message appears: `Writing login information to the keychain failed with error 'Unknown or unsupported transport “disabled” for address “disabled:”'.`
![image](https://user-images.githubusercontent.com/5366828/113366472-4b398300-9327-11eb-8958-df865c72201b.png)

Does this issue occur when all extensions are disabled?: Yes

## Before this gets closed as a duplicate of other issues:
- yes, I have `libsecret` and `gnome-keyring` installed, as other issues describe
- yes, I have a keyring and it is set as default
- other applications are able to use the keyring successfully

**Comment by maosid** [comment]
I had the same problem before, in my case i need to add `dbus-launch` in` .xinitrc` to run dbus-session

- ~/.xinitrc

```
# init keyring
eval $(/usr/bin/gnome-keyring-daemon --start --components=gpg,pkcs11,secrets,ssh)
# export keyring
export GNOME_KEYRING_CONTROL GNOME_KEYRING_PID GPG_AGENT_INFO SSH_AUTH_SOCK

...

exec dbus-launch dwm
```

**Comment by Echika** [comment]
Unable to Authorize Github what do I do?

**Comment by SergioRius** [comment]
> Upstream issue about this [atom/node-keytar#132](https://github.com/atom/node-keytar/issues/132)
> 
> I've updated the troubleshooting guide to point to the instructions from @maosid. Thanks!

And those who doesn't use a gnome based system, like kubuntu?

**Comment by moghaazi** [comment]
> > Upstream issue about this [atom/node-keytar#132](https://github.com/atom/node-keytar/issues/132)
> > I've updated the troubleshooting guide to point to the instructions from @maosid. Thanks!
> 
> And those who doesn't use a gnome based system, like kubuntu?

Yup, I use Arch with Sway window manager, and this solution didn't fix the issue, I also tried this solution #92972 didn't fix it either

**Comment by bdombro** [comment]
I am having the same/similar issue on Osx:

```
Writing login information to the keychain failed with error 'UNIX[No such file or directory]'.
```

**Comment by innoplusdevs** [comment]
@maosid, in which part of the file should I put this code? because there is already code in this file.

**Comment by innoplusdevs** [comment]
@maosid  what would be the code for KDE?

**Comment by nan0tube** [comment]
> > > Upstream issue about this [atom/node-keytar#132](https://github.com/atom/node-keytar/issues/132)
> > > I've updated the troubleshooting guide to point to the instructions from @maosid. Thanks!
> > 
> > 
> > And those who doesn't use a gnome based system, like kubuntu?
> 
> Yup, I use Arch with Sway window manager, and this solution didn't fix the issue, I also tried this solution #92972 didn't fix it either

I use Gentoo, had the same issue and resolved it by launching sway via `dbus-run-session sway`. Hope this helps someone.

**Comment by nelsonov** [comment]
As a kludge, it's also possible to simply do:

`dbus-launch code`

The dbus-launch command sets some environment variables for the process that it spawns.  If you are spawning a window manager in .xinitrc (ie `dbus-launch dwm`) then the entire GUI session will have those environment variables set.

If you run just `dbus-launch` at the cli, then you see the variables it is trying to set, ie:
```
> dbus-launch
DBUS_SESSION_BUS_ADDRESS=unix:abstract=/tmp/dbus-sqROfzYKp1,guid=e9e634929ddff76d580cde8c618a3862
DBUS_SESSION_BUS_PID=30017
```
If you have launched your window manager from dbus-launch in .xinitrc or wherever, then you should be able to see those variables already existing by:
> env | grep DBUS_SESSION_BUS

For those familiar with ssh-agent, this is the same concept.

**Comment by julianperezpesce** [comment]
I use Manjaro KDE, I can't fix the annoying problem, any help?

**Comment by yehia1ali** [comment]
@julianperezpesce
I had the same problem with manjaro plasma
The problem was solved by :
Add the following lines to your ~/.xinitrc:
`# see https://unix.stackexchange.com/a/295652/332452
source /etc/X11/xinit/xinitrc.d/50-systemd-user.sh

# see https://wiki.archlinux.org/title/GNOME/Keyring#xinitrc
eval $(/usr/bin/gnome-keyring-daemon --start)
export SSH_AUTH_SOCK

# see https://github.com/NixOS/nixpkgs/issues/14966#issuecomment-520083836
mkdir -p "$HOME"/.local/share/keyrings`
then Command:
`sudo pacman -S gnome-keyring libsecret libgnome-keyring`
install seahorse if dosen't exist .
Now restart vscode - he is work with me

**Comment by MahmoudAshraf25** [comment]
Anyone on kde still facing the problem, simply install gnome-keyring

I'm on arch so `sudo pacman -S gnome-keyring` did the job

**Comment by itsbek** [comment]
I'm facing the same problem, but the error message is different and that's why haven't tried any solutions for now...
![here it is](https://user-images.githubusercontent.com/65237794/150644474-8b8e3b03-ce33-4d27-ab28-fb2c0150ee85.png)
any thoughts?

**Comment by MoRadwan74** [comment]
> I'm facing the same problem, but the error message is different and that's why haven't tried any solutions for now... ![here it is](https://user-images.githubusercontent.com/65237794/150644474-8b8e3b03-ce33-4d27-ab28-fb2c0150ee85.png) any thoughts?

Me too!

**Comment by itsbek** [comment]
> > I'm facing the same problem, but the error message is different and that's why haven't tried any solutions for now... ![here it is](https://user-images.githubusercontent.com/65237794/150644474-8b8e3b03-ce33-4d27-ab28-fb2c0150ee85.png) any thoughts?
> 
> Me too!

for me, surprisingly [these](https://stackoverflow.com/a/64306522/13784922) steps worked out

**Comment by yashank09** [comment]
Still facing the initial issue on artix(arch) even after installing `gnome-keyring & libsecret` and adding the above lines to my .xinitrc. 
Any other solutions?

**Comment by ntrf** [comment]
This error should not pop up at all if there is no settings sync configured to begin with. Don't tell me "But we need to check somehow". You should add a boolean into the setting disabling keychain access. If that means no settings sync -- that's fine.

**Comment by yashank09** [comment]
> This error should not pop up at all if there is no settings sync configured to begin with. Don't tell me "But we need to check somehow". You should add a boolean into the setting disabling keychain access. If that means no settings sync -- that's fine.

Wait that literally solves nothing though. I clearly have credentials I would like to save but none of the methods above solve it. I have isolated the issue and it seems dbus is having a hard time to start on boot. If I fix it, ill post something here.

**Comment by LeoniePhiline** [comment]
How can I use KDE Wallet (kwalletd) instead? Gnome keyring does not integrate well into a KDE workstation.

**Comment by YerongLi2** [comment]
> I had the same problem before, in my case i need to add `dbus-launch` in` .xinitrc` to run dbus-session
> 
>     * ~/.xinitrc
> 
> 
> ```
> # init keyring
> eval $(/usr/bin/gnome-keyring-daemon --start --components=gpg,pkcs11,secrets,ssh)
> # export keyring
> export GNOME_KEYRING_CONTROL GNOME_KEYRING_PID GPG_AGENT_INFO SSH_AUTH_SOCK
> 
> ...
> 
> exec dbus-launch dwm
> ```

I am on Debian 11-updates/bookworm version (Antix Linux)
Adding throse lines ~/.xinitrc and reboot my laptop does not help me either.

Also I tested 

     eval $(/usr/bin/gnome-keyring-daemon --start --components=gpg,pkcs11,secrets,ssh)

And 

     eval $(/usr/bin/gnome-keyring-daemon --start)
 
Either of them gives error message:
     
     ** Message: 01:17:28.652: couldn't connect to control socket at: /run/user/1000/keyring/control: Connection refused

Anyone met similar issues?

**Comment by wiltonlazary** [comment]
Please, stop been opinionated how about Linux and keep vscode things on vscode code space, stop trying to use desktop services because Linux users are very diversified and everything was OK until now.

**Comment by wiltonlazary** [comment]
Please, stop being opinionated how about Linux and keep vscode things on vscode code space, stop trying to use desktop services because Linux users are very diversified and everything was OK until now.

**Comment by LeoniePhiline** [comment]
It's an improvement if they do it right and support all popular desktops and protocols.

**Comment by yashank09** [comment]
> > I had the same problem before, in my case i need to add `dbus-launch` in` .xinitrc` to run dbus-session
> > ```
> > * ~/.xinitrc
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > # init keyring
> > eval $(/usr/bin/gnome-keyring-daemon --start --components=gpg,pkcs11,secrets,ssh)
> > # export keyring
> > export GNOME_KEYRING_CONTROL GNOME_KEYRING_PID GPG_AGENT_INFO SSH_AUTH_SOCK
> > 
> > ...
> > 
> > exec dbus-launch dwm
> > ```
> 
> I am on Debian 11-updates/bookworm version (Antix Linux) Adding throse lines ~/.xinitrc and reboot my laptop does not help me either.
> 
> Also I tested
> 
> ```
>  eval $(/usr/bin/gnome-keyring-daemon --start --components=gpg,pkcs11,secrets,ssh)
> ```
> 
> And
> 
> ```
>  eval $(/usr/bin/gnome-keyring-daemon --start)
> ```
> 
> Either of them gives error message:
> 
> ```
>  ** Message: 01:31:07.785: couldn't access control socket: /run/user/1000/keyring/control: No such file or directory
> ```
> 
> Anyone met similar issues?

Yep same here on Artix s6 machine. I believe this is a problem with our setups and not VS Code itself, so we shouldnt be hoping for any over the top solution from the team.

**Comment by YerongLi** [comment]
> > > I had the same problem before, in my case i need to add `dbus-launch` in` .xinitrc` to run dbus-session
> > > ```
> > > * ~/.xinitrc
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > ```
> > > # init keyring
> > > eval $(/usr/bin/gnome-keyring-daemon --start --components=gpg,pkcs11,secrets,ssh)
> > > # export keyring
> > > export GNOME_KEYRING_CONTROL GNOME_KEYRING_PID GPG_AGENT_INFO SSH_AUTH_SOCK
> > > 
> > > ...
> > > 
> > > exec dbus-launch dwm
> > > ```
> > 
> > 
> > I am on Debian 11-updates/bookworm version (Antix Linux) Adding throse lines ~/.xinitrc and reboot my laptop does not help me either.
> > Also I tested
> > ```
> >  eval $(/usr/bin/gnome-keyring-daemon --start --components=gpg,pkcs11,secrets,ssh)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > And
> > ```
> >  eval $(/usr/bin/gnome-keyring-daemon --start)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Either of them gives error message:
> > ```
> >  ** Message: 01:31:07.785: couldn't access control socket: /run/user/1000/keyring/control: No such file or directory
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Anyone met similar issues?
> 
> Yep same here on Artix s6 machine. I believe this is a problem with our setups and not VS Code itself, so we shouldnt be hoping for any over the top solution from the team.


Yes, I guess the best we can do is try whether we can fix this through discussions. The recent progress I met is I tried to add **dbus-launch**

    dbus-launch /usr/share/code/code --unity-launch
    
 The error message becomes 

     Writing login information to the keychain failed with error 'Error calling StartServiceByName for org.freedesktop.secrets: Timeout was reached'.

![screenshot](https://user-images.githubusercontent.com/13112023/164106027-6cc21bac-233e-4640-ae2f-e8d5157c45b5.jpg)


Any idea how to fix this ?

**Comment by ghos5y** [comment]
I Think i found a "workaround", i opened VS Code, got the error message and exited. I reopened VS Code and waited for the "Unlock the default keyring" thing, typed in the password and it worked. Or im just delusional and they fixed it.

**Comment by FadhilAF** [comment]
> > > > I had the same problem before, in my case i need to add `dbus-launch` in` .xinitrc` to run dbus-session
> > > > ```
> > > > * ~/.xinitrc
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > ```
> > > > # init keyring
> > > > eval $(/usr/bin/gnome-keyring-daemon --start --components=gpg,pkcs11,secrets,ssh)
> > > > # export keyring
> > > > export GNOME_KEYRING_CONTROL GNOME_KEYRING_PID GPG_AGENT_INFO SSH_AUTH_SOCK
> > > > 
> > > > ...
> > > > 
> > > > exec dbus-launch dwm
> > > > ```
> > > 
> > > 
> > > I am on Debian 11-updates/bookworm version (Antix Linux) Adding throse lines ~/.xinitrc and reboot my laptop does not help me either.
> > > Also I tested
> > > ```
> > >  eval $(/usr/bin/gnome-keyring-daemon --start --components=gpg,pkcs11,secrets,ssh)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > And
> > > ```
> > >  eval $(/usr/bin/gnome-keyring-daemon --start)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Either of them gives error message:
> > > ```
> > >  ** Message: 01:31:07.785: couldn't access control socket: /run/user/1000/keyring/control: No such file or directory
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Anyone met similar issues?
> > 
> > 
> > Yep same here on Artix s6 machine. I believe this is a problem with our setups and not VS Code itself, so we shouldnt be hoping for any over the top solution from the team.
> 
> Yes, I guess the best we can do is try whether we can fix this through discussions. The recent progress I met is I tried to add **dbus-launch**
> 
> ```
> dbus-launch /usr/share/code/code --unity-launch
> ```
> 
> The error message becomes
> 
> ```
>  Writing login information to the keychain failed with error 'Error calling StartServiceByName for org.freedesktop.secrets: Timeout was reached'.
> ```
> 
> ![screenshot](https://user-images.githubusercontent.com/13112023/164612621-872d3be4-4de3-4cb4-81dc-3a3e2f1492d3.jpg)
> 
> Any idea how to fix this ?

I have the same error message as you

**Comment by jokeroy2022** [comment]
# see https://unix.stackexchange.com/a/295652/332452
source /etc/X11/xinit/xinitrc.d/50-systemd-user.sh

# see https://wiki.archlinux.org/title/GNOME/Keyring#xinitrc
eval $(/usr/bin/gnome-keyring-daemon --start)
export SSH_AUTH_SOCK

# see https://github.com/NixOS/nixpkgs/issues/14966#issuecomment-520083836
mkdir -p "$HOME"/.local/share/keyrings

**Comment by kuberlog** [comment]
I'm getting `Error calling StartServiceByName for org.freedesktop.secrets: Timeout was reached`

Gentoo
awesomwm

I've added the stuff to xinitrc

**Comment by KafCoppelia** [comment]
Any idea about KDE Wallet instead of gnome-keyring?

**Comment by kabalage** [comment]
I'm also getting the error `Writing login information to the keychain failed with error 'Error calling StartServiceByName for org.freedesktop.secrets: Timeout was reached'`

I use arch with i3 and I have all the above configured.

**Comment by amanos1022** [comment]
The issue for me seems to be conflicting "ssh agents", `ssh-agent` and `gnome-keyring-daemon`. 

In order to avoid entering my ssh key's pw for every remote git command, I use this command to start my window manager session (debian + i3) from the console:
```bash
eval `ssh-agent -s`; ssh-add; startx; # this would prompt me for my ssh key pair password and start i3wm
```
My understanding of the issue is that VSCode requires `gnome-keyring` and that `ssh-agent` does the same thing or something (or points to something that's the same thing, not an expert in all this)  and somehow interferes with VSCode and `gnome-keyring` interfacing. 

So something like this should fix the issue (or at least point you in the right direction).

```bash
export $(gnome-keyring-daemon --start) # Haven't figured out why we run this then kill it later
ssh-agent -k # kill whatever ssh-agent is (could be kwallet or debian keyring?)

killall gnome-keyring-daemon # Not sure why but I have to kill here. The command is restarted by VSCode it seems.
export $(dbus-launch) # I actually have this in my .bashrc
ssh-add # enter your password
```
Pretty sure I had to do the following as well to setup the gnome keyring
```bash
# backup old keyring
mv ~/.local/share/keyrings/Default_keyring.keyring ~/.local/share/keyrings/Default_keyring.keyring.backup
```
Sorry in advanced for my lack of understanding surrounding `ssh-agent` and `gnome-keyring-daemon`, ssh agents in general, or other terminology and/or bad info. This is just what has worked for me. I spent a lot of time and tried a lot of things so I might be missing a step. Hopefully this helps someone

**Comment by holms** [comment]
Absolutely nothing helped from this ticket. How could this ticket be closed if nothing works. I've tried all possible solutions in here. Even when opening seahorse there's no section "Passwords". All commands launched fine, but vscode doesn't see anything when trying to login for Copilot.

**Comment by stickyburn** [comment]
As labeled in the ticket, this is out of scope from officiial support so linux related forms might help better.

**Comment by holms** [comment]
> As labeled in the ticket, this is out of scope from officiial support so linux related forms might help better. 

That's sad. Why GitHub copilot extension is using keyrings at all? Why supporting only gnome users? Who made this decision?

**Comment by kabalage** [comment]
I struggled a couple of hours to get it working on my arch setup, but I always ran into some new error, so I gave up... I have to log in with GitHub every time a VSCode window opens (every window), because it can't save the credentials. Pretty annoying.

**Comment by holms** [comment]
> I struggled a couple of hours to get it working on my arch setup, but I always ran into some new error, so I gave up... I have to log in with GitHub every time a VSCode window opens (every window), because it can't save the credentials. Pretty annoying.

I've tried so many things.. and now it works. And I have no idea why. Seahorse didn't show any keychain vaults at all, and then it started to show it ¯\_₍⸍⸌̣ʷ̣̫⸍̣⸌₎_/¯ 

I think this might happened because I've unlocked gpg key when I've committed to git repo. Because nothing else been changed at all in my setup. Although I'm not sure what does my personal gpg key has anything to do with default gnome-keyring vault

**Comment by kuberlog** [comment]
I personally gave up using copilot because of this.

**Comment by holms** [comment]
> I personally gave up using copilot because of this. 

Well done Microsoft 🤣

**Comment by ElYaiko** [comment]
> I personally gave up using copilot because of this.

Same :/

**Comment by holms** [comment]
Ok I think I've got a fix.

* `rm -rf ~/.local/share/keyrings/*`
* launch `seahorse`
* On top-left (before "Passwords and Keys") hit `+` and choose "Password keyring", name it "Login" and _DO_NOT_ set password. 
* Launch vscode, push sign-in button "Default keyring" will be created
* Pop up will appear with asking to enter new password, do not enter anything, push create

**Comment by tuananh** [comment]
i'm on arch / sway and nothing works. any idea?
got gnome-keyring, libgnome-keyring and libsecret installed.
